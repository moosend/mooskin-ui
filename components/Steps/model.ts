import { IBoxComponentProps } from '../Box/model';

export interface IStepsComponentProps extends IBoxComponentProps {
	/** number of the active item or items */
	activeItem?: number | string;

	/** onclick callback when the item is clicked */
	onClickStep?: (e: React.MouseEvent<HTMLElement>, value?: number | string) => void;

	children?: React.ReactElement<IStepCommonComponentProps> | Array<React.ReactElement<IStepCommonComponentProps>>;
}

export interface IStepCommonComponentProps extends IBoxComponentProps {
	/** whether the current item is active */
	active?: boolean;
}

export interface IStepHeaderComponentProps extends IStepCommonComponentProps {
	/** whether this step is disabled */
	disabled?: boolean;
}

export interface IStepComponentProps extends IStepCommonComponentProps {
	/** identifier for togglin the Stepbed item */
	activeId?: string | number;
}
