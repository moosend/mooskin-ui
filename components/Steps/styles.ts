import styled from 'styled-components';

// Models
import { IStepCommonComponentProps, IStepComponentProps, IStepHeaderComponentProps, IStepsComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

import variables from '../_utils/globals/variables';

export const StyledSteps = styled(Box)<IStepsComponentProps>`
	display: flex;
	flex-direction: column;
`;

StyledSteps.displayName = 'StyledSteps';

export const StyledStep = styled(Box)<IStepComponentProps>``;

StyledStep.displayName = 'StyledStep';

export const StyledStepHeader = styled(Box)<IStepHeaderComponentProps>`
	border-bottom: ${(props) => (props.active ? 'solid 2px #293346' : 'solid 2px #9d9d9d')};
	color: ${(props) =>
		props.active
			? `${props.palette?.fontColors.text || variables.fontColors.text}`
			: `${props.palette?.fontColors.medgray1 || variables.fontColors.medgray1}`};
	width: 200px;
	height: 40px;
	font-size: 14px;
	font-weight: 500;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	display: flex;
	justify-content: center;
	align-items: center;
	cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
	opacity: ${(props) => (props.disabled ? 0.5 : 1)};
	position: relative;
`;

StyledStepHeader.displayName = 'StyledStepHeader';

export const StyledStepContent = styled(Box)<IStepCommonComponentProps>``;

StyledStepContent.displayName = 'StyledStepContent';

export const StyledStepArrow = styled(Box)<IStepCommonComponentProps>`
	font-family: 'Mooskin Icons';
	position: absolute;
	top: 10px;
	right: -5px;
	color: inherit;
	font-size: 12px;
`;

StyledStepArrow.displayName = 'StyledStepArrow';
