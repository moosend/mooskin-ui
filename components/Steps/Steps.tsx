import * as React from 'react';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Models
import { IStepCommonComponentProps, IStepComponentProps, IStepHeaderComponentProps, IStepsComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

// Styled Components
import { StyledStep, StyledStepArrow, StyledStepContent, StyledStepHeader, StyledSteps } from './styles';

/**
 * Steps
 */
export const Steps: React.FC<IStepsComponentProps> = withMooskinContext((props) => {
	const batchClickHandler = (
		e: React.MouseEvent<HTMLElement>,
		activeId?: string | number,
		callback?: (e: React.MouseEvent<HTMLElement>) => void
	) => {
		props.onClickStep && props.onClickStep(e, activeId);
		callback && callback(e);
	};

	const getActiveItem = (activeId?: string | number) => {
		if (props.activeItem && Array.isArray(props.activeItem)) {
			return (props.activeItem as any).includes(activeId);
		}
		return props.activeItem === activeId;
	};

	const assignChildren = (tabIndex: number, tabsLength: number, children?: any, activeId?: string | number, active?: boolean) => {
		let header: React.ReactElement<IStepCommonComponentProps> | undefined;
		let content: React.ReactElement<IStepCommonComponentProps> | undefined;

		const hasArrow = tabIndex < tabsLength - 1;

		React.Children.forEach(children, (child: any) => {
			if (React.isValidElement<IStepHeaderComponentProps>(child) && child.type === StepHeader) {
				header = React.cloneElement(child, {
					active,
					children: (
						<>
							{child.props.children}
							{hasArrow && <StepArrow children="keyboard_arrow_right" />}
						</>
					),
					onClick: (e) => (!child.props.disabled ? batchClickHandler(e, activeId, child.props.onClick) : undefined),
				} as IStepHeaderComponentProps);
			}

			if (React.isValidElement<IStepCommonComponentProps>(child) && child.type === StepContent) {
				content = active ? React.cloneElement(child, { active } as IStepCommonComponentProps) : undefined;
			}
		});

		return { header, content };
	};

	const recurseChildren = (children: any): any => {
		if (!children) {
			return null;
		}

		const headers: Array<React.ReactElement<IStepCommonComponentProps> | null> = [];
		const contents: Array<React.ReactElement<IStepCommonComponentProps> | null> = [];

		React.Children.forEach(children, (child, i) => {
			if (React.isValidElement<IStepComponentProps>(child) && child.type === Step) {
				const active = child.props.active ? child.props.active : getActiveItem(child.props.activeId);
				const { content, header } = assignChildren(i, children.length, child.props.children, child.props.activeId, active);
				header && headers.push(React.cloneElement(header, { key: i }));
				content && contents.push(React.cloneElement(content, { key: i }));
			}
		});

		return (
			<>
				<Box d="flex">{headers}</Box>
				{contents}
			</>
		);
	};

	return <StyledSteps {...props} children={recurseChildren(props.children)} />;
});

Steps.defaultProps = {
	className: '',
	style: {},
};

Steps.displayName = 'Steps';

/**
 * Step
 */
export const Step: React.FC<IStepComponentProps> = withMooskinContext((props) => {
	return <StyledStep {...props} />;
});

Step.defaultProps = {
	className: '',
	style: {},
};

Step.displayName = 'Step';

/**
 * StepHeader
 */
export const StepHeader: React.FC<IStepHeaderComponentProps> = withMooskinContext((props) => {
	return <StyledStepHeader {...props} />;
});

StepHeader.defaultProps = {
	className: '',
	style: {},
};

StepHeader.displayName = 'StepHeader';

/**
 * StepContent
 */
export const StepContent: React.FC<IStepCommonComponentProps> = withMooskinContext((props) => {
	return <StyledStepContent {...props} />;
});

StepContent.defaultProps = {
	className: '',
	style: {},
};

StepContent.displayName = 'StepContent';

/**
 * StepArrow
 */
const StepArrow: React.FC<IStepCommonComponentProps> = withMooskinContext((props) => {
	return <StyledStepArrow {...props} />;
});

StepArrow.defaultProps = {
	className: '',
	style: {},
};

StepArrow.displayName = 'StepArrow';
