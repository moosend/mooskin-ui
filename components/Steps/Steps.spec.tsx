import * as React from 'react';
import { Step, StepContent, StepHeader, Steps } from './Steps';

import { mount } from 'enzyme';

describe('Pagination', () => {
	test('renders correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Steps activeItem={3} onClickStep={func}>
				<Step activeId={1}>
					<StepHeader>{`Item: 1`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 1`}</div>
					</StepContent>
				</Step>
				<Step activeId={2}>
					<StepHeader>{`Item: 2`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 2`}</div>
					</StepContent>
				</Step>
				<Step activeId={3}>
					<StepHeader>{`Item: 3`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 3`}</div>
					</StepContent>
				</Step>
				<Step activeId={4}>
					<StepHeader>{`Item: 4`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 4`}</div>
					</StepContent>
				</Step>
				<Step activeId={5}>
					<StepHeader>{`Item: 5`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 5`}</div>
					</StepContent>
				</Step>
				<Step activeId={6}>
					<StepHeader>{`Item: 6`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 6`}</div>
					</StepContent>
				</Step>
			</Steps>
		);

		expect(tree).toMatchSnapshot();
	});

	test('renders active Step content correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Steps activeItem={3} onClickStep={func}>
				<Step activeId={1}>
					<StepHeader>{`Item: 1`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 1`}</div>
					</StepContent>
				</Step>
				<Step activeId={2}>
					<StepHeader>{`Item: 2`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 2`}</div>
					</StepContent>
				</Step>
				<Step activeId={3}>
					<StepHeader>{`Item: 3`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 3`}</div>
					</StepContent>
				</Step>
				<Step activeId={4}>
					<StepHeader>{`Item: 4`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 4`}</div>
					</StepContent>
				</Step>
				<Step activeId={5}>
					<StepHeader>{`Item: 5`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 5`}</div>
					</StepContent>
				</Step>
				<Step activeId={6}>
					<StepHeader>{`Item: 6`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 6`}</div>
					</StepContent>
				</Step>
			</Steps>
		);

		expect(tree.find('StyledStepContent').length).toEqual(1);
		expect(tree.find('StyledStepContent').text()).toEqual('Content for item with index: 3');
	});

	test('renders active Step content correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Steps activeItem={3} onClickStep={func}>
				<Step activeId={1}>
					<StepHeader>{`Item: 1`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 1`}</div>
					</StepContent>
				</Step>
				<Step activeId={2}>
					<StepHeader>{`Item: 2`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 2`}</div>
					</StepContent>
				</Step>
				<Step activeId={3}>
					<StepHeader>{`Item: 3`}</StepHeader>
					<StepContent>
						<div>{`Content for item with index: 3`}</div>
					</StepContent>
				</Step>
			</Steps>
		);

		tree.find(StepHeader).first().simulate('click');
		expect(func).toHaveBeenCalled();
	});
});
