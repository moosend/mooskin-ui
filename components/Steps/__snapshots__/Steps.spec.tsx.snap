// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps renders correctly 1`] = `
<div
  className="steps "
>
  <div
    className="stepsContainer"
  >
    <Step
      active={true}
      id="1"
      title="Step 1"
    >
      Content 1
    </Step>
    <Step
      id="2"
      title="Step 2"
    >
      Content 2
    </Step>
    <Step
      id="3"
      title="Step 3"
    >
      Content 3
    </Step>
  </div>
  <div>
    <StepBody
      active={true}
      content="Content 1"
      key=".0"
    />
    <StepBody
      content="Content 2"
      key=".1"
    />
    <StepBody
      content="Content 3"
      key=".2"
    />
  </div>
</div>
`;
